src/gc.c

```diff
@@ -149,10 +149,10 @@ static void finalize_object(arraylist_t *list, jl_value_t *o,
     if (oldlen == len)
         return;
     if (need_sync) {
-        if (__unlikely(jl_atomic_compare_exchange(&list->len,
-                                                  oldlen, len) != oldlen)) {
-            memset(items[len], 0, (oldlen - len) * sizeof(void*));
-        }
+        jl_atomic_compare_exchange(&list->len, oldlen, len);
+        // The memset needs to be unconditional since the thread might have
+        // already read the length.
+        memset(&items[len], 0, (oldlen - len) * sizeof(void*));
```

Comment:
`len` is a local variable and it is only "published" (i.e. released) with the `cmpxchg` after all the content modifications are done so it is important that we update the `len` but the `memset` should be before the `cmpxchg` when the `cmpxchg` succeeds and is fine to be on either side if `cmpxchg` fails.

* For both `finalize_object` and `gc_add_finalizer_`, the synchronization happens through the `arraylist_t::len` field and only happens at the beginning and the end of the function.
* There can be at most two threads operating on the same list, one in `gc_add_finalizer_` and one in `finalize_object`
* `gc_add_finalizer_` only increase the length and mutate content after the length it reads whereas `finalize_object` only decrease the length and mutate the content before the length it reads, and all content mutation happens in between the acquire of the `len` and release (including the `seq_cst` ordering of the `cmpxchg`) of the `len`. This way, the mutation on a thread cannot be observed on another thread before the release of the `len` (and will be consistently observed after that).
* Finally it is important that the `len` increment always succeeds so `gc_add_finalizer_` uses a release store and the `finalize_object` use `cmpxchg`. The decrement doesn't have to succeed since it can clear the memory instead.